### PR TITLE
[eas-cli] add --auto flag to publish

### PR DESCRIPTION
**Pull request by jkhales**
# Checklist

- [x] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary.
- [x] I've tagged the changelog entry with `[EAS BUILD API]` if it's a part of a breaking change to EAS Build API (only possible when updating `@expo/eas-build-job` package).

# Why

We want a flag to automatically set the branch name and update message to the git branch and commit message.

Also made command create a branch if it doesn't already exist.

# Test Plan

Tested in demo repo.

### CODE REVIEW

File: packages/eas-cli/src/commands/branch/publish.ts [hunk 1]
```diff
@@ -318,12 +362,29 @@ export default class BranchPublish extends Command {
...
+  let actorName: string;
+  switch (actor?.__typename) {
+    case 'User': {
+      actorName = (actor as Pick<User, 'username' | 'id'>).username;
+      break;
+    }
+    case 'Robot': {
+      const { firstName, id } = actor as Pick<Robot, 'firstName' | 'id'>;
+      actorName = firstName ?? `robot: ${id.slice(0, 4)}...`;
+      break;
```
Comment (jkhales) [inline packages/eas-cli/src/commands/branch/publish.ts:381]: This is not strictly related to the `--auto` addition. Noticed we missed a case when fixing up the typing. Curious if this will fix that robot issue @jonsamp was having?

File: packages/eas-cli/src/commands/branch/publish.ts [hunk 2]
```diff
@@ -171,11 +212,10 @@ export default class BranchPublish extends Command {
         })
       );
     }
-    assert(name, 'branch name must be specified.');
-
-    const { id: branchId, updates } = await viewUpdateBranchAsync({
+    assert(branchName, 'branch name must be specified.');
+    const { id: branchId, updates } = await ensureBranchExists({
```
Comment (jkhales) [inline packages/eas-cli/src/commands/branch/publish.ts:216]: No longer throw if branch doesn't exist. Create it instead.

File: packages/eas-cli/src/commands/branch/publish.ts [hunk 3]
```diff
@@ -107,13 +137,19 @@ export default class BranchPublish extends Command {
       description: `return a json with the new update group.`,
       default: false,
     }),
+    auto: flags.boolean({
```
Comment (wkozyra95) [inline packages/eas-cli/src/commands/branch/publish.ts:140]: maybe consider sth more descriptive, meaning of the auto in context of publishing is not really obvious

`--auto-branch`/`--auto-message`
`--managed`
or use `--non-interactive` flag and if values are not provided via `--message`/`--branch` us those defaults

Also if you are considering adding updates config to eas.json, options like this would be a good candidate for it

Comment (jkhales) [reply packages/eas-cli/src/commands/branch/publish.ts:140]: The issue with `--non-interactive` is that is not clear that this means "apply an opinionated default" instead of "throw if essential args are not provided"

`--auto-branch/--auto-message` are nice, but maybe too verbose?

> Also if you are considering adding updates config to eas.json, options like this would be a good candidate for it

Interesting idea, I'm a bit hesitant to engage there right now as it is in such flux that things tend to get reverted. Maybe once things get more established?

@jonsamp curious what you think about these other flag suggestions.

File: packages/eas-cli/src/commands/branch/publish.ts [hunk 4]
```diff
@@ -152,14 +188,19 @@ export default class BranchPublish extends Command {
     }
     const projectId = await getProjectIdAsync(exp);
 
-    if (!name) {
+    if (!branchName && autoFlag) {
+      branchName =
+        (await vcs.getBranchNameAsync()) || `branch-${Math.random().toString(36).substr(2, 4)}`;
```
Comment (jonsamp) [inline packages/eas-cli/src/commands/branch/publish.ts:193]: We should create another task to make the `getBranchNameAsync` work inside GitHub Actions. The command this runs is `git rev-parse --abbrev-ref HEAD`, which works locally, but results in this error during a GitHub Action:

Screen Shot 2021-08-06 at 12 06 27 AM

This is not a blocker for this PR, and this should generally not be a problem for CI, since we can pass in the branch name/message programmatically. But it'd be great to run `eas branch:publish --auto` locally, and then write the same command for the CI action.

File: packages/eas-cli/src/commands/branch/publish.ts [hunk 5]
```diff
@@ -171,11 +212,10 @@ export default class BranchPublish extends Command {
         })
       );
     }
-    assert(name, 'branch name must be specified.');
-
-    const { id: branchId, updates } = await viewUpdateBranchAsync({
+    assert(branchName, 'branch name must be specified.');
```
Comment (jonsamp) [inline packages/eas-cli/src/commands/branch/publish.ts:215]: I think we could move this `assert` inside the conditional above it. It's a check to make sure that the user chose a branch via an interactive prompt, so it seems like it belongs in there, since the other flow will get the git branch or create a random branch.

File: packages/eas-cli/src/commands/branch/publish.ts [hunk 6]
```diff
@@ -271,6 +311,10 @@ export default class BranchPublish extends Command {
       }
     }
 
+    if (!message && autoFlag) {
+      message = (await vcs.getLastCommitMessageAsync())?.trim();
```
Comment (jonsamp) [inline packages/eas-cli/src/commands/branch/publish.ts:315]: Similarly here, this fails when run on CI. It runs `git --no-pager log -1 --pretty=%B`, but since there's no git repo in an action, we'll see:

Screen Shot 2021-08-06 at 12 15 09 AM

Again, no need to update that logic in this PR, but let's make some tasks.

File: packages/eas-cli/src/commands/branch/publish.ts (hunk 1 above)
Comment (jonsamp) [reply packages/eas-cli/src/commands/branch/publish.ts:381]: Yeah, this should do it! Thanks for adding this along the way

File: packages/eas-cli/src/commands/branch/view.ts [hunk 7]
```diff
@@ -131,10 +110,15 @@ export default class BranchView extends Command {
       }));
     }
 
-    const UpdateBranch = await viewUpdateBranchAsync({
+    // return data.app?.byId.updateBranchByName;
```
Comment (jonsamp) [inline packages/eas-cli/src/commands/branch/view.ts:113]:
```suggestion
```

**Review by jonsamp — APPROVED** [review]
Left some questions and suggestions but this code looks good to me. Excited to use this feature!

File: packages/eas-cli/src/commands/branch/publish.ts (hunk 4 above)
Comment (wkozyra95) [reply packages/eas-cli/src/commands/branch/publish.ts:193]: .git repository should exist after checkout, if it doesn't then I think it's bug in github action.

Seems like it might be cause by old git version in the image 
https://github.com/actions/checkout/issues/335
https://github.com/actions/checkout/issues/126

**Review by wkozyra95 — APPROVED** [review]

File: packages/eas-cli/src/commands/branch/publish.ts (hunk 5 above)
Comment (jkhales) [reply packages/eas-cli/src/commands/branch/publish.ts:215]: It's also meant as a defensive check in case someone changes the code where it's set automatically.

Seems costless to have it serve both purposes.

File: packages/eas-cli/src/commands/branch/view.ts (hunk 7 above)
Comment (jkhales) [reply packages/eas-cli/src/commands/branch/view.ts:113]: whoops, thanks.

File: packages/eas-cli/src/commands/branch/publish.ts (hunk 4 above)
Comment (jkhales) [reply packages/eas-cli/src/commands/branch/publish.ts:193]: @jonsamp let me know if @wkozyra95 suggestions for the github action don't work out.

File: packages/eas-cli/src/commands/branch/publish.ts (hunk 5 above)
Comment (jkhales) [reply packages/eas-cli/src/commands/branch/publish.ts:215]: Well, maybe you are right.

File: packages/eas-cli/src/commands/branch/publish.ts (hunk 3 above)
Comment (jkhales) [reply packages/eas-cli/src/commands/branch/publish.ts:140]: Thanks for the discussion, will go with `--auto` for now!
